ReactWindows/ReactNative.Shared/UIManager/AccessibleAutomationPeer.cs

```diff
@@ -0,0 +1,104 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using ReactNative.UIManager.Events;
+#if WINDOWS_UWP
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Automation.Peers;
+using Windows.UI.Xaml.Automation.Provider;
+#else
+using System.Windows;
+using System.Windows.Automation.Peers;
+using System.Windows.Automation.Provider;
+#endif
+
+namespace ReactNative.UIManager
+{
+    /// <summary>
+    /// Automation peer for a FrameworkElement that implements IAccessible.
+    /// </summary>
+    /// <typeparam name="T">Type of owner of the automation peer.</typeparam>
+    public sealed class AccessibleAutomationPeer<T> : FrameworkElementAutomationPeer, IInvokeProvider
+        where T : FrameworkElement, IAccessible
+    {
+        /// <summary>
+        /// Hides base UIElement Owner to provide stronger-typed T Owner.
+        /// </summary>
+        private new T Owner => (T)base.Owner;
+
+        /// <inheritdoc />
+        public AccessibleAutomationPeer(T owner)
+            : base(owner)
+        {
+        }
+
+        /// <inheritdoc />
+        protected override IList<AutomationPeer> GetChildrenCore()
+        {
+            bool shouldHideChildren =
+                Owner.ImportantForAccessibility == ImportantForAccessibility.Yes
+                || Owner.ImportantForAccessibility == ImportantForAccessibility.NoHideDescendants;
+            return shouldHideChildren ? null : base.GetChildrenCore();
+        }
+
+        /// <inheritdoc />
+        protected override bool IsContentElementCore()
+        {
+            return Owner.ImportantForAccessibility == ImportantForAccessibility.Yes;
+        }
+
+        /// <inheritdoc />
+        protected override bool IsControlElementCore()
+        {
+            return Owner.ImportantForAccessibility == ImportantForAccessibility.Yes;
+        }
+
+        /// <inheritdoc />
+        protected override AutomationControlType GetAutomationControlTypeCore()
+        {
+            if (Owner.AccessibilityTraits?.Contains(AccessibilityTrait.Button) == true)
+            {
+                return AutomationControlType.Button;
```

Comment:
>return AutomationControlType.Button; [](start = 16, length = 36)

What happens if an element is a Button and then switches to something else or vice versa? Does that scenario work properly?

Reply:
no. this is what github rnw #1562 is about. I have TODO in code referencing the #1562 in AccessibleCanvas where implementations of importantForAccessibility and accessibilityTraits are.

---
In reply to: [158562054](https://github.com/reseul/react-native-windows/pull/1#discussion_r158562054) [](ancestors = 158562054)